### PR TITLE
fix(model): make Model.validate use object under validation as context by default

### LINKED ISSUE #10346: Validate doesn't work with conditional required fields
**Do you want to request a *feature* or report a *bug*?**
Bug.

**What is the current behavior?**
Conditional `required` fields don't throw when using `Model.validate()`.

**If the current behavior is a bug, please provide the steps to reproduce.**

```js
const mongoose = require('mongoose');
const assert = require('assert');

run().catch(console.error);

async function run () {
  const userSchema = new mongoose.Schema({
    name: { type: String, required: true },
    age: { type: Number, required () {return this.name === 'John';} }
  });

  const User = mongoose.model('User', userSchema);

  const err = await User.validate({ name: 'John' }).then(() => null, err => err);
  assert.ok(err);
}
```

**What is the expected behavior?**
Expects method to throw a validaiton error path `age` is requried.

**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**
Node.js: 14.15.5
Mongoose: 5.12.13
MongoDB:  4.4.3

### PR DESCRIPTION
**Pull request by AbdelrahmanHafez**
fixes #10346

### CODE REVIEW

File: lib/model.js [hunk 1]
```diff
@@ -4062,9 +4062,9 @@ Model.aggregate = function aggregate(pipeline, callback) {
  */
 
 Model.validate = function validate(obj, pathsToValidate, context, callback) {
-  if (context === void 0 && obj instanceof Model) {
-    // For convenience, if we're validating a document, make `context` default to
-    // the model so users don't have to always pass `context`, re: gh-10132.
+  if (context === void 0) {
```
Comment (vkarpov15) [inline lib/model.js:4065]: I think we should make this only happen if the user doesn't specify `context`. So either `arguments.length < 3`, or `arguments.length === 3 && typeof arguments[2] === 'function'`.

**Review by vkarpov15 — COMMENTED** [review]
I agree that this is the right way to go. Using `undefined` by default is a bit messy. Maybe we should release this in 5.13 though, might be a bit heavy for a patch release.

**Comment by AbdelrahmanHafez** [comment]
Sounds good to me.

**Comment by AbdelrahmanHafez** [comment]
@vkarpov15 
Regarding the `context` change, and how we're trying to guess what parameter is what. There's this pattern [RORO](https://www.freecodecamp.org/news/elegant-patterns-in-modern-javascript-roro-be01e7669cbd/), receive an object, return an object.
It makes function parameters similar to python's named parameters, so that you won't have to guess anymore. We've been using those everywhere around our codebases and never went back.

We've also implemented a layer over mongoose to make mongoose's API comply with that pattern.
That pattern has been specifically helpful in testing, because stubbing chainable APIs can be really difficult.
```js
    Comment.dal.find({
      filter: { userId: user._id },
      skip,
      limit,
      sort: sortBy,
      populate: { path: 'author', select: 'name username email' },
      lean: true
    }),
```

Does converting mongoose APIs to take objects with named parameters like that seem like an idea you'd pursue someday? Should I open an issue so we can discuss it further?

**Comment by vkarpov15** [comment]
Yeah using objects to simulate [named parameters in JavaScript](https://masteringjs.io/tutorials/fundamentals/parameters) is a neat pattern, and I'm not opposed to potentially supporting it in the future.

**Review by miladezzat — APPROVED** [review]

**Review by vkarpov15 — APPROVED** [review]
Thanks :+1:

**Comment by AbdelrahmanHafez** [comment]
Happy to help 👍

**Pull request merged**